contracts/cosmwasm-vm/cw-ibc-core/src/ics02_client/handler.rs

```diff
@@ -164,9 +211,70 @@ impl<'a> IbcClient for CwIbcCoreContext<'a> {
     fn upgrade_client(
         &self,
         deps: DepsMut,
-        message: ibc::core::ics02_client::msgs::upgrade_client::MsgUpgradeClient,
-    ) {
-        todo!()
+        info: MessageInfo,
+        message: MsgUpgradeClient,
+    ) -> Result<Response, ContractError> {
+        let old_client_state = self.client_state(deps.as_ref().storage, &message.client_id)?;
+
+        //Check Client Frozen
+        if old_client_state.is_frozen() {
+            return Err(ContractError::IbcClientError {
+                error: ClientError::ClientFrozen {
+                    client_id: message.client_id,
+                },
+            });
+        }
+
+        let old_consensus_state = self.consensus_state(
+            deps.as_ref().storage,
+            &message.client_id,
+            &old_client_state.latest_height(),
+        )?;
+
+        let now = self.host_timestamp()?;
+        let duration = now
+            .duration_since(&old_consensus_state.timestamp())
+            .ok_or_else(|| ClientError::InvalidConsensusStateTimestamp {
+                time1: old_consensus_state.timestamp(),
+                time2: now,
+            })
+            .map_err(|error| ContractError::IbcClientError { error })?;
+
+        // Check if the latest consensus state is within the trust period.
+        if old_client_state.expired(duration) {
+            return Err(ContractError::IbcClientError {
+                error: ClientError::HeaderNotWithinTrustPeriod {
+                    latest_time: old_consensus_state.timestamp(),
+                    update_time: now,
+                },
+            });
+        };
+
+        // Validate the upgraded client state and consensus state and verify proofs against the root
+
+        let wasm_exec_message = LightClientMessage::UpgradeClient {
+            upgraded_client_state: message.client_state.value,
+            upgraded_consensus_state: message.consensus_state.value,
+            proof_upgrade_client: to_vec(&message.proof_upgrade_client).unwrap(),
+            proof_upgrade_consensus_state: to_vec(&message.proof_upgrade_consensus_state).unwrap(),
+        };
+
```

Comment:
I am not sure if relay would be able to provide proofs for client state and consensus state in case of icon chain. Need to discuss this with @AntonAndell . I have not implemented this on light client side as well.

Reply:
Client state and consensus state should be able to be proven, but client upgrades are not part of the current scope

Reply:
In that case, this will not be exposed to the entry point of the contract. But implementation will be in place. When it is required we can update the logic and enable it as an entry point. is that fine?